sdk/lib/ksanitize/ubsan.c

```diff
@@ -0,0 +1,1828 @@
+/* $NetBSD: ubsan.c,v 1.10 2020/03/08 21:35:03 kamil Exp $ */
+
+/*-
+ * Copyright (c) 2018 The NetBSD Foundation, Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
+ * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * The micro UBSan implementation for the userland (uUBSan) and kernel (kUBSan).
+ * The uBSSan versions is suitable for inclusion into libc or used standalone
+ * with ATF tests.
+ *
+ * This file due to long symbol names generated by a compiler during the
+ * instrumentation process does not follow the KNF style with 80-column limit.
+ */
+
+#ifdef __REACTOS__
+
+#include <inttypes.h>
+#include <limits.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <sys/types.h>
+
+#ifndef _NTSYSTEM_
+#define _NTSYSTEM_
+#endif
+
+#ifndef __NTDLL__
+#define _NTDLLBUILD_
+#endif
+
+#include <ntdef.h>
+#include <ntddk.h>
+
+#define __printflike(n,m) __attribute__((format(printf,n,m)))
+#define __unreachable __builtin_unreachable
+#define __arraycount(a) (sizeof(a)/sizeof(a[0]))
+
+#ifndef __BIT
+#define __BIT(__n) \
+  (((size_t)(__n) >= CHAR_BIT * sizeof(size_t)) ? 0 : \
+  ((size_t)1 << (size_t)((__n) & (CHAR_BIT * sizeof(size_t) - 1))))
+#endif
+
+#ifndef __LOWEST_SET_BIT
+/* find least significant bit that is set */
+#define __LOWEST_SET_BIT(__mask) ((((__mask) - 1) & (__mask)) ^ (__mask))
+#define __SHIFTOUT(__x, __mask) (((__x) & (__mask)) / __LOWEST_SET_BIT(__mask))
+#define __SHIFTIN(__x, __mask) ((__x) * __LOWEST_SET_BIT(__mask))
+#define __SHIFTOUT_MASK(__mask) __SHIFTOUT((__mask), (__mask))
+#endif
+
+#ifdef LIBSAN_KMODE
+#include <ndk/kdfuncs.h>
+#elif LIBSAN_W32K
+
+#undef ASSERT
+#define ASSERT(x)
```

Comment:
What's wrong with `ASSERT` here? Missing `RtlAssert`? Can we instead use something like
```
#define ASSERT(exp) ((!(exp)) ? DbgRaiseAssertionFailure(), FALSE) : TRUE))
```
That should work anywhere.